File: fisica-computacional/examen_1/src/ejercicio_3_2.c
/*
 *============================================================================
 * Program: Prime Number Tester (C Implementation)
 * Purpose: Determine if a given number is prime using trial division
 * Author:  Luis Fernando Núñez Rangel
 * Date:    2025
 *
 * Description:
 *   Tests primality of an integer using the trial division method.
 *   This is the C99 implementation parallel to the Fortran version.
 *
 * Algorithm:
 *   - Numbers 1, 2, 3 are considered prime (base cases)
 *   - For n > 3: test divisibility by all integers from 2 to n-1
 *   - Return 1 (true) if prime, 0 (false) if composite
 *
 * Compilation:
 *   gcc -std=c99 ejercicio_3_2.c -o ejercicio_3_c
 *
 * Usage:
 *   ./ejercicio_3_c
 *   Enter an integer when prompted
 *============================================================================
 */

#include <stdio.h>

/* Function prototype */
int is_p(int a);

/*
 * Main Program
 * Reads an integer from user and determines if it's prime
 */
int main(int argc, char **argv) {
    int number = 0;

    /* Input: Read number from user */
    printf("Enter a number: ");
    scanf("%d", &number);

    /* Output: Display primality test result */
    /* Ternary operator: condition ? true_value : false_value */
    printf("Is prime number: %s\n", is_p(number) ? "true" : "false");

    return 0;
}

/*
 *============================================================================
 * Function: is_p
 * Purpose:  Test if a number is prime using trial division
 *
 * Parameters:
 *   a - Integer to test for primality
 *
 * Returns:
 *   1 if number is prime
 *   0 if number is composite
 *
 * Algorithm:
 *   1. Base case: numbers 1-3 are prime
 *   2. Trial division: test all divisors from 2 to a-1
 *   3. If any divisor divides evenly (a % i == 0), return 0
 *   4. If no divisors found, return 1
 *
 * Complexity:
 *   Time: O(n) where n is the input number
 *   Space: O(1)
 *
 * Optimization Note:
 *   Could be improved to O(√n) by testing only up to square root of a
 *============================================================================
 */
int is_p(int a) {
    /* Base case: numbers 1, 2, 3 are prime */
    if (a >= 1 && a <= 3) {
        return 1;
    }

    /* Trial division: test all potential divisors */
    for (int i = 2; i < a; i++) {
        if (a % i == 0) {
            /* Found a divisor - number is composite */
            return 0;
        }
    }
    
    /* No divisors found - number is prime */
    return 1;
}

File: fisica-computacional/examen_1/src/ejercicio_3_2.f90
!===============================================================================
! Program: es_primo
! Purpose: Test if a given number is prime
! Author:  Luis Fernando Núñez Rangel
! Date:    2025
!
! Description:
!   Determines whether an input integer is prime using trial division method.
!   A prime number is divisible only by 1 and itself.
!
! Algorithm:
!   - Numbers 1, 2, 3 are considered prime (base cases)
!   - For n > 3: test divisibility by all integers from 2 to n-1
!   - If any divisor found, number is composite
!   - Otherwise, number is prime
!
! Complexity:
!   Time: O(n) - Could be optimized to O(√n) by testing only up to sqrt(n)
!   Space: O(1)
!
! Compilation:
!   gfortran ejercicio_3_2.f90 -o ejercicio_3_2
!
! Usage:
!   ./ejercicio_3_2
!   Enter an integer when prompted
!===============================================================================

program es_primo
    implicit none
    
    ! Variable declarations
    integer :: a          ! Number to test for primality
    integer :: resultado  ! Result flag: 1 = prime, 0 = composite

    !---------------------------------------------------------------------------
    ! Input: Read number from user
    !---------------------------------------------------------------------------
    print *, "Enter a number: "
    read *, a

    !---------------------------------------------------------------------------
    ! Test primality using subroutine
    !---------------------------------------------------------------------------
    call is_p(a, resultado)

    !---------------------------------------------------------------------------
    ! Output: Display result
    !---------------------------------------------------------------------------
    if (resultado == 1) then
        print *, "true"   ! Number is prime
    else 
        print *, "false"  ! Number is composite
    end if

end program es_primo

!===============================================================================
! Subroutine: is_p
! Purpose: Determine if a number is prime using trial division
!
! Arguments:
!   a         [in]  - Integer to test for primality
!   resultado [out] - Result flag: 1 if prime, 0 if composite
!
! Algorithm:
!   1. Handle base cases (1, 2, 3 are prime)
!   2. Test divisibility by all integers from 2 to a-1
!   3. If divisible by any number, it's composite (resultado = 0)
!   4. If no divisors found, it's prime (resultado = 1)
!
! Note:
!   This is a basic implementation. For large numbers, use optimized
!   algorithms like Miller-Rabin or AKS primality test.
!===============================================================================
subroutine is_p(a, resultado)
    implicit none
    
    ! Argument declarations
    integer, intent(in) :: a
    integer, intent(out) :: resultado
    
    ! Local variables
    integer :: i  ! Loop counter for divisibility testing

    !---------------------------------------------------------------------------
    ! Base case: small numbers (1, 2, 3) are prime
    !---------------------------------------------------------------------------
    if (a >= 1 .AND. a <= 3) then 
        resultado = 1
        return 
    end if

    !---------------------------------------------------------------------------
    ! Trial division: test all potential divisors from 2 to a-1
    !---------------------------------------------------------------------------
    do i = 2, a - 1
        if (mod(a, i) == 0) then
            ! Found a divisor - number is composite
            resultado = 0
            return
        end if 
    end do

    !---------------------------------------------------------------------------
    ! No divisors found - number is prime
    !---------------------------------------------------------------------------
    resultado = 1

end subroutine is_p
